text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,111 @@
+- Start Date: 2023-02-01
+- RFC Type: feature
+- RFC PR: n/a
+- RFC Status: draft
+
+# Summary
+
+This RFC proposes introducing a centralized schema repository for the Kafka topics in use at Sentry. It provides mappings between the Kafka topics we are using to the encoding format and schemas for messages on that topic.
+
+Topic registration and schema validation of messages will be optional for the foreseeable future.
+
+# Motivation
+
+Kafka is increasingly used as the message bus between services at Sentry. A Kafka topic and its’ schema not owned by any one service, but is part of the contract between services.
+
+As Sentry grows, the number of event types and topics in use is exploding. We are marching further towards increasingly distributed ownership of data by product and infrastructure engineering teams. We have already seen an increase in the number of incidents related with invalid data and schema issues at Sentry. It is reasonable to expect this trend to continue as the number of topics, data types, teams and engineers contributing to Sentry grows.
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
+- explicitly enumerate ownership of schemas
+- make it easier for consumers to identify and reject bad messages without pausing the whole pipeline
+
+# Supporting Data
+
+Over the last quarter, we have seen many incidents related to schema disagreements at Sentry. Examples include the following: (note: inc numbers are internal to Sentry)
+
+- Post process forwarder (INC-218)
+- Snuba’s transactions consumer (INC-210)
+- Super big consumers (INC-220)
+- Replays consumer (INC-281)
+
+Many of these incidents are P1. Since messages are in order, an invalid message often halts consumers and requires manual intervention. This is disruptive to both Sentry engineers and our users (it takes us much longer to recover).
+
+# Options Considered
+
+### **Option A (preferred): Publish a library**
+
+Schemas will be made available as a library for multiple programming languages (Python and Rust to start).
+
+**Client usage example:**
+
+```python
+from sentry_kafka_schemas import get_schema
+
+topic = "events"
+schema = get_schema(topic)
+# => {
+# 	"type": "json"
+# 	"schema": {
+# 		"$schema": "http://json-schema.org/draft-07/schema#",
+# 		"type": "object",
+# 		...
+# 	}
+# }
+
+# Get a specific version of the schema
+v1_schema = get_schema(topic, version=1)
+```
+
+**How schemas are stored:**
+Topic data will be stored as yaml. For example:
+
+```
+topic: events
+schemas:
+    - version: 1
+    type: json
+    resource: events.schema.json
+    - version: 2
+    type: avro
```

Comment:
We also have grpc/protobufs floating around in our codebase. Do we need json, avro, and protobufs?

Reply:
Good question. We will most likely focus on supporting json to start. The reason for this is not technical, just that we are focussing on golden paths and don't necessarily want to encourage too many different encodings and schema formats at Sentry. But this type field will be here from the start, and the process to add other formats should be pretty straightforward.

Reply:
we can probably hammer msgpack payloads into a json-shaped hole if we have to. grpc is another world

Reply:
Agreed. We are using msgpack in Sentry in at least one place so we may need to add it sooner rather than later. I think msgpack can still use the jsonschema though, just the decoder would be different.